engine.py

```diff
@@ -108,6 +116,57 @@ def post_app_init(self):
             except:
                 traceback.print_exc()
 
+    def validate_context(self, tank, context):
```

Comment:
I think there are benefits to keep it here. More easy to maintain, more easy to understand quickly.